hadoop-hdds/docs/content/design/container-reconciliation.md

```diff
@@ -0,0 +1,246 @@
+        1. If the chunk or block is not present in our container, save it in memory as needing to be imported.
+        2. If the chunk or block is not marked `healthy` in our merkle tree (known to be corrupted by the scanner), save it in memory as needing to be reconciled.
+        3. If chunk checksums differ but our chunk is marked `healthy`, ignore this chunk.
+            - If the chunk is corrupted in the other replica or got corrupted while we were reading it, we should not import it.
+            - If the chunk is corrupted on one of the replicas but the `healthy` flag has not been updated yet to reflect this, it will be resolved when the scanner reaches the container again.
+            - If the same chunk is truly `healthy` in both replicas but the checksums do not match, there has been an irrecoverable error on the write path where different data was written to each replica but they both passed checksum validation.
+        4. Datanode calls existing `readChunk` API on the other datanode for each chunk that was selected for repair during the tree scan.
+        5. Datanode builds/patches/adds to block file.
+            - Use existing read/write chunk APIs for this. Read/write checksum validation is built in to these already.
+    4. Datanode recomputes container merkle tree from new chunks and existing checksums that were not repaired.
+        - By recomputing the tree after reconciling with each replica, it can exit the loop if the trees from all other replicas match.
+
+### Concurrent Events
+
+These may happen to the container during reconciliation and must remain consistent during reconciliation. Note that we cannot depend on SCM coordination to shelter datanodes from other commands after SCM has instructed them to reconcile because datanode command queues may get backed up at various points and process commands in a different order, or long after they were initially issued.
+
+#### On Client read of container under reconciliation
+
+- No action required.
+
+This read could be an end client or another datanode asking for data to repair its replica. The result will only be inconsistent if the chunk/block is being repaired. In this case the data is already corrupt and the client will get a read checksum failure either way. Even if all datanodes are reconciling the same container at the same time, this should not cause an availability problem because there has to be one correct replica that others are repairing from that the client can also read.
+
+#### Replicate command for a container being reconciled
+
+- Allow replication to proceed with best effort, possibly with a warning log.
+
+There may be cases where a container is critically under-replicated and we need a good copy of the data as soon as possible. Meanwhile reconciliation could remained blocked for an unrelated reason. We can provide a way to pause/unpause reconciliation between chunk writes to do reasonably consistent container exports.
+
+#### On block delete
```

Comment:
Its not clear to me how this mechanism can detect "lost deletes" where a block should be deleted from all replicas, but someone the block got left behind on one or more of them.

If the MT ignores the deleted blocks so that when they are deleted the container hash does not change, then is there any efficient way to detect lost deletes?

Lost block deletes are something we want to address, and it is something we should design into this solution too.

Say, on delete, we update a flag in the MT, but do not adjust the container hash. We cannot detect lost deletes at SCM based on the root hash alone. Therefore nothing would trigger reconciliation unless there is an additional corruption in a replica to trigger it.

I agree that the container hash should not be changed due to a delete, but I think we many need something else to highlight lost delete problems - eg a deleted hash (combination of checksums of deleted block) and possibly a deleted block list or deleted block MT. I am further unsure if a MT could help find a lost delete, as they are not suited for proof of membership / non membership AFAIK. Sorted block lists may be need to be compared.

Reply:
The Merkle Tree records blocks that have been deleted. The hash of the container does not change but the Merkle Tree's metadata for the block changes.